lib/experimental/html/class-wp-html-walker.php

```diff
@@ -0,0 +1,1277 @@
+
+		// Find tag name's end in the updated markup.
+		$markup_updated_up_to_a_tag_name_end = $this->updated_html . substr( $this->html, $this->updated_bytes, $this->tag_name_ends_at - $this->updated_bytes );
+		$updated_tag_name_ends_at            = strlen( $markup_updated_up_to_a_tag_name_end );
+		$tag_name_length                     = $this->tag_name_ends_at - $this->tag_name_starts_at;
+		$updated_tag_name_starts_at          = $updated_tag_name_ends_at - $tag_name_length;
+
+		// Apply attributes updates.
+		$this->updated_html  = $markup_updated_up_to_a_tag_name_end;
+		$this->updated_bytes = $this->tag_name_ends_at;
+		$this->class_name_updates_to_attributes_updates();
+		$this->apply_attributes_updates();
+
+		// Replace $this->html with the updated markup.
+		$this->html = $this->updated_html . substr( $this->html, $this->updated_bytes );
+
+		// Rewind the walker to the tag name's end.
+		$this->tag_name_starts_at = $updated_tag_name_starts_at;
+		$this->tag_name_ends_at   = $updated_tag_name_ends_at;
+		$this->parsed_bytes       = $this->tag_name_ends_at;
+
+		// Restore the previous version of the updated_html as we are not finished with the current_tag yet.
+		$this->updated_html  = $markup_updated_up_to_a_tag_name_end;
+		$this->updated_bytes = $updated_tag_name_ends_at;
+
+		// Parse the attributes in the updated markup.
+		$this->attributes = array();
+		$this->parse_tag_opener_attributes();
+
+		return $this->html;
+	}
+
+	/**
+	 * Prepares tag search criteria from input interface.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param array|string $query {
+	 *     Which tag name to find, having which class.
+	 *
+	 *     @type string|null $tag_name     Which tag to find, or `null` for "any tag."
+	 *     @type string|null $class_name   Tag must contain this class name to match.
+	 * }
+	 */
+	private function parse_query( $query ) {
+		if ( null !== $query && $query === $this->last_query ) {
+			return;
+		}
+
+		$this->last_query          = $query;
+		$this->sought_tag_name     = null;
+		$this->sought_class_name   = null;
+		$this->sought_match_offset = 1;
+
+		// A single string value means "find the tag of this name".
+		if ( is_string( $query ) ) {
+			$this->sought_tag_name = $query;
+			return;
+		}
+
+		// If not using the string interface we have to pass an associative array.
+		if ( ! is_array( $query ) ) {
+			return;
+		}
+
+		if ( isset( $query['tag_name'] ) && is_string( $query['tag_name'] ) ) {
+			$this->sought_tag_name = $query['tag_name'];
+		}
+
+		if ( isset( $query['class_name'] ) && is_string( $query['class_name'] ) ) {
+			$this->sought_class_name = $query['class_name'];
+		}
+
+		if ( isset( $query['match_offset'] ) && is_int( $query['match_offset'] ) && 0 < $query['match_offset'] ) {
+			$this->sought_match_offset = $query['match_offset'];
+		}
+	}
+
+
+	/**
+	 * Checks whether a given tag and its attributes match the search criteria.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @return boolean
+	 */
+	private function matches() {
```

Comment:
before seeing your comment I was wondering about adding a comment to the final `return true`. it's at least been my practice in these situations to avoid returning _even though we can_ because it breaks the ladder of conditions and I want to try and avoid accidentally skipping a condition added at a later time. for example, we started with a third condition - "the tag must have this `data-` attribute" - and then removed it.

so before removing it we had this

```php
if ( $needs_tag_match ) {
	…
	return false;
}

if ( $needs_class_match ) {
	…
	return false;
}

if ( $needs_data_attribute_match ) {
	…
	return false;
}

return true;
```

that last `true` means "none of the pre-conditions rejected the match" and although we can change that structure I feel like it's worth keeping in, but now I'm sure it clearly warrants a comment.

coincidentally, because I didn't see this review yet, I pushed a commit that removed the nested check for needing a class name but not having a class field.

> No tag name === "matches".

these are currently designed as search criteria, so no tag name means "a tag with no restrictions on the name" just as no class name means "a tag with no restrictions on which classes it contains."

to visit every tag in a document, we do this

```php
while ( $walker->next_tag() ) { // alternatively next_tag( [] )
	// process tag here
}
```

this also means we can't return true before checking the class name requirements.

> This private method would be called when $this->sought_tag_name is null.

we might still need to match a class name. for example, we might want to find the first tag having a given CSS class, which is common in the theming code and where many blocks have a user-controllable tag name.

```php
if ( $walker->next_tag( [ 'class_name' => 'fullwidth' ] ) ) {
	// process full-width element here
}
```

Reply:
> Correct me if wrong, but it looks like we can return early on two conditions to reduce nesting and, in the latter case, processing.
> 
> ```
> if ( null = $this->sought_tag_name ) {
>     return true;
> }
> 
> if ( ! isset( $this->attributes['class'] ) ) {
>     return false;
> }
> ```
> 
> Also: The first returns `true`, as in "Yes, the given tag and its attributes match the search criteria", and I'm not quite sure why.

This doesn't capture the logic of the `matches()` method. In pseudocode, it's more like this:

```php
if ( $should_match_tag_name ) {
    // Match, return false on mismatch
}

if ( $should_match_attributes ) {
    // Match, return false on mismatch
}

return true;
```

It cannot be reduced to:

```php
if ( ! $should_match_tag_name ) {
    return true;
}

if ( ! $has_class_name ) {
    return false;
}

```

Reply:
I'll resolve this for now as it takes up a lot of vertical space. Feel free to unresolve if needed, @costdev

Reply:
I've just realised that I had mistakenly read the conditions as both being `if ( null !== $this->sought_tag_name ) {`, and missed that the second was actually `if ( null !== $this->sought_class_name ) {` which is why I thought it was unnecessary repetition. My bad on that! :sweat_smile:

An early return is still valid if both are `null`, and would reduce nesting, but I understand that this is a matter of preference here.

Thanks to both for the replies to explain regarding returning `true` when either of these properties is `null` and for clarifying the logic of the method!